### ISSUE TITLE
🐞 existing services break in 0.15

### ISSUE THREAD

**Issue by levlaz** [dagger/dagger#9190]
### What is the issue?

I have a sample go application (https://github.com/levlaz/snippetbox) that installs mariadb (https://daggerverse.dev/mod/github.com/levlaz/daggerverse/mariadb@cea1668da940b45864116049bd20087855c8c787) from the daggerverse and mounts it as a service binding. 

THe mariadb module is targeting 0.11.6 and the snippetbox module is targeting   0.13.16

This is no longer working in 0.15 with the following error: 

```
│ ✔ .withEnvVariable(name: "CACHEBUSTER", value: "2024-12-12 22:56:27.108879878 +0000 UTC m=+0.053134793"): Container! 0.0s
│ ✘ .withExec(args: ["sh", "-c", "mysql -h db -u root < internal/db/init.sql"]): Container! 0.1s
│ ┃ mysql: Deprecated program name. It will be removed in a future release, use '
│ ┃ /bin/mariadb' instead                                                        
│ ┃ ERROR 2026 (HY000): TLS/SSL error: self-signed certificate                   
│ ! process "sh -c mysql -h db -u root < internal/db/init.sql" did not complete successfully: exit code: 1
│ ○ .withExec(args: ["sh", "-c", "mysql -h db -u root snippetbox < internal/db/seed.sql"]): Container! 0.0s
│ ○ .withExec(args: ["go", "run", "./cmd/web", "--dsn", "web:pass@tcp(db)/snippetbox?parseTime=true"]): Container! 0.0s
✘ .terminal: Container! 4.9s
! failed to start service: solve mount /: process "sh -c mysql -h db -u root < internal/db/init.sql" did not complete successfully: exit code: 1
```

I suspect its not able to get the service to work due to the new API change, but I was under the impression that this change should not break existing pipelines like this. 

### Dagger version

dagger v0.15.0 (registry.dagger.io/engine:v0.15.0) darwin/arm64

### Steps to reproduce

dagger -m github.com/levlaz/snippetbox call server

### Log output

Here's a trace: https://dagger.cloud/levs-test-org/traces/f06d326bb79ffd99fb7afdc647b58f21

**Comment by rajatjindal** [comment]
Hi @levlaz 

Thank you for reporting this issue. 

I just tried this with v0.14.0 and got the same error:

```
mysql: Deprecated program name. It will be removed in a future release, use '/usr/bin/mariadb' instead
ERROR 2026 (HY000): TLS/SSL error: self-signed certificate
```

(trace: https://v3.dagger.cloud/rajatjindal/traces/7a4d4678fc04d42c8071fe1d03c4e4bb?listen=b913daa7677b6b45&listen=5d710621d7bdd220)

once I added `--skip-tls` to mysql commands, it passed through that error. (Kindly note that it still gets stuck at `withExec` to start the golang server, that I am trying to understand why and if that was always the case)

could you please confirm if that makes sense or did I missed something when trying to reproduce the problem?

**Comment by rajatjindal** [comment]
I think I know what might be happening here. when calling `dagger container up`, it calls internally the `asService` api, which seems like is going to the new implementation even when the module is at version `< v0.15.0`. 

@jedevc is that expected?

We may have to add a legacy vs non-legacy version of `dagger container up` for this.

**Comment by jedevc** [comment]
Hm, yes. `container.Up` should also be versioned.

**Comment by jedevc** [comment]
It actually might be possible to do something like this - instead of needing to handle a legacy function:

```go
	var svc dagql.Instance[*core.Service]
	err := s.srv.Select(ctx, ctr, &svc,
		dagql.Selector{
			Field: "asService",
			View:  s.srv.View,
		},
	)
```

But that's just a guess, we need a test for this as well.

**Comment by jedevc** [comment]
> pass the server view if we use `s.srv.View`

Is this not what we want?

**Comment by jedevc** [comment]
The server is created *from* the module source, so I would expect this to work.

**Comment by jedevc** [comment]
`hack/dev dagger query --doc ./modules/compatcheck/introspection.graphql` this is definitely expected - the `dagger` command here would be using `v0.15.2`.

If you call this in the context of a module - it should give the right result though.

**Comment by rajatjindal** [comment]
I confirmed by adding logging that the view is set to v0.14.0 (from dagger.json) when calling the function that returns a container. BUT when we call `up` on that container, the version seen on server side is `v0.15.2` (the current engine version). 

I am trying to identify how to fix this.

**Comment by jedevc** [comment]
@rajatjindal what's the command you're running to see this? Are you doing `dagger call server up`?

**Comment by jedevc** [comment]
Okay okay, I synced with @rajatjindal on this, and we worked out next steps.

So - the *reason* this breaks is this:
- `server` returns a `Container`. We then call `up` on it (in the context of the cli, which has been upgraded to v0.15).
- I think the right thing to do here is to change `server` to return a `Service` instead of a `Container` - it should be a ready-to-run `Service`. In this case, the `Service` conversion would happen completely as intended with module compat.
- But when returning a `Container` and converting it to a `Service` *once* we've crossed this boundary just can't really work.

Aside from the fixes we'll make below - I think it's worth updating snippetbox to return a `Service` from this function.

> [!NOTE]
>
> Why can't we just use the API versioned from the module in the CLI? Why can't we just have `Container.up` here use the v0.14.0 API from the module instead of the v0.15.0 API from the CLI?
>
> This is a deliberate design decision we made when doing module compat - without this, it's *really* weird. Suppose we make a breaking change to an API in `withExec`. Now, if that applies on the CLI as well, as an external caller, I need to modify my call to match the API version of the module I'm calling - which isn't easy to guess or know. This would be really confusing, so we just present *one* version constantly - the CLI *always* gets it's version of the API (v0.15.0 in this case).

That said, the current state is horribly difficult to migrate to, so here's what we're going to do:
- `Container.up` isn't properly versioned - you'd only hit this if you called `Container.up` in a module, so that's not exactly the issue here, but we should still fix it. @jedevc will fix this.
- `Container.up` should have *all* the args from `Container.asService`, and *all* the args from `Service.up` - this would let you do `dagger call server up --args=...`. @jedevc will fix this.
- `Container.up` and `Container.asService` should fail if no args are provided anywhere. This'll fix the infinite hanging problem where the last `WithExec` will be run as default, and also catch this case. @rajatjindal will fix this.
	- This would still break the example in snippetbox that you're using @levlaz - but it would provide a significantly improved error message.
	- We might also extend this to version `WithDefaultArgs` - if snippetbox's `server` method used `WithDefaultArgs`, we wouldn't trigger this error - but if we add a versioned `WithDefaultArgs`, we could be able to detect this case, and error out. If we do this, @jedevc will fix this.

I *think* this is the best we can do, keeping the API design that we've done - sadly, I don't think there's a feasible way to get the snippetbox module working with v0.15.0 as-is, with no changes to either the module or the `call` command that invokes it. I think this is probably a miss from us, we should have caught this significantly earlier in the process - might be worth a retro to work out how to not do this again.

The aim is to try and fix this ASAP, and get it into v0.15.2 sometime early this week.

**Comment by jpadams** [comment]
Possible summary note for Dagger 0.15+ regarding how we expect things to work:

# Dagger, ENTRYPOINT, CMD/DefaultArgs behavior

Dagger does not use the container `ENTRYPOINT` by default in`AsService` or `WithExec` functions, but uses `CMD`/`defaultArgs` by default. This allows the `CMD`/`defaultArgs` to be overridden by args  provided to `AsService` or `WithExec`.

## How to handle entrypoint scripts/commands in `ENTRYPOINT`:

### `AsService`:
Since many popular container images for long-running services like dbs and webservers currently specify a script for initialization in the `ENTRYPOINT` args array,  we advocate combining the `ENTRYPOINT` + `CMD` args  and setting that as the new `CMD`/`defaultArgs` to make the behavior most explicit and predicable at runtime and on publish. For example:

`dagger core container from --address postgres without-entrypoint with-default-args --args docker-entrypoint.sh,postgres with-exposed-port --port 5432 with-env-variable --name POSTGRES_PASSWORD --value postgres as-service up`

It's easy to discover the values for the `ENTRYPOINT` and `CMD`/`defaultArgs` by running a command like:
`dagger core container from --address postgres`

original
```
_type: Container
defaultArgs:
    - postgres
entrypoint:
    - docker-entrypoint.sh
mounts: []
platform: linux/arm64
user: ""
workdir: ""
```
then examine the preferred container:
`dagger core container from --address postgres without-entrypoint with-default-args --args docker-entrypoint.sh,postgres with-exposed-port --port 5432 with-env-variable --name POSTGRES_PASSWORD --value postgres`
```
_type: Container
defaultArgs:
    - docker-entrypoint.sh
    - postgres
entrypoint: []
mounts: []
platform: linux/arm64
user: ""
workdir: ""
```

### `WithExec`:
Some container images may similarly have a command like `git` in the `ENTRYPOINT`. We advocate for moving commands like that to the args of the `WithExec` where you execute that command. For example `alpine/git` has `git` in the `ENTRYPOINT` and ``--help` in the `CMD`/`defaultArgs`:

`dagger core container from --address alpine/git without-entrypoint without-default-args with-exec --args git,version stdout`

As above:
`dagger core container from --address alpine/git`
```
_type: Container
defaultArgs:
    - --help
entrypoint:
    - git
mounts: []
platform: linux/arm64
user: ""
workdir: /git
```

### Using existing `ENTRYPOINT`:
Alternatively, you can  use `UseEntrypoint` as a convenience to use the default `ENTRYPOINT` specified in the image together with either the `CMD`/`defaultArgs` or args you specify in the`AsService` or `WithExec` function calls.

**Comment by marcosnils** [comment]
Thx for the writeup Jeremy, it's super clear.

> Dagger does not use the container `ENTRYPOINT` by default in`AsService` or `WithExec` functions, but uses `CMD`/`defaultArgs` by default. This allows the `CMD`/`defaultArgs` to be overridden by args provided to `AsService` or `WithExec`.

Should we expand this a bit and briefly explain why?

**Comment by marcosnils** [comment]
@jpadams  let's not forget to update your last comment based on the latest interactions with @shykes :pray:

**Comment by jpadams** [comment]
**tl;dr:** For calls to `Up` or `AsService` on a `Container` without `args` and with NO mutations/overrides to the `ENTRYPOINT+CMD`/`defaultArgs`, execute `ENTRYPOINT+CMD` on service start. (see **In practice**…`AsService` below)  
<img src="https://github.com/user-attachments/assets/2ed8ad9f-2b5b-4e7c-9df0-aa9a63bf8ce5" width="200" />

**The Dagger philosophy:**  
Dagger combines both ***build*** capabilities of `Dockerfile` / `docker` `build` and ***run*** capabilities of `docker` `run` (interactive and service/daemon mode) and `docker` `compose`, while adding additional capabilities not available in those tools. Our desire is for a unified experience with a DX that is intuitive for both experienced users of Docker and Kubernetes as well as users new to containers and pipelines.

For the best containerized pipelines-as-code experience, users should not rely upon the OCI container image’s `ENTRYPOINT` or `CMD`, but rather explicitly state their desired arguments in the `WithExec` or `AsService` functions as `args`. In this way, the arguments are kept “close” to the point of execution and users avoid any confusion that could come about from relying on the implicit `ENTRYPOINT+CMD` behavior used by default in Docker and Kubernetes or the `ENTRYPOINT+args` behavior in `docker` `run`. There is an exception to this rule to accommodate the ecosystem of service containers (e.g. database, webserver, cache, proxy) that use `ENTRYPOINT` to execute an `entrypoint.sh` script to perform essential service configuration (see `AsService` below).

*Note: In order to have compatibility with the Docker and Kubernetes's ecosystems, users have full control over setting the `ENTRYPOINT` and `CMD` when building container images for consumption outside of Dagger.*

**In practice, we should support and advise users to follow these guidelines:**  
`WithExec` \- Always prefer explicit use of `WithExec(args)`. These `args` will override the `CMD`/`defaultArgs` (which should not be relied on either, in almost all cases). Since `ENTRYPOINT` is ignored by default, `args` becomes the primary source of truth.  
**Build case:** When building container images, `WithExec` is similar to the `Dockerfile` `RUN` (example \[1\], \[2\]).  
**Run case:** When “running” container images by using functions such as `stdout`, `stderr`, `sync`, or `terminal`, `WithExec` is similar to `docker` `run` where we want to obtain the output of a last command or interact with a container after a command has been run (example \[3\]). 

\[1\] `dagger shell -s -c 'container | from alpine | with-exec echo,first | stdout'`  
`first`

\[2\] `dagger shell -s -c 'container | from alpine | with-exec echo,first | with-exec echo,last | stdout'`  
`last`

\[3\] `dagger shell -c 'container | from alpine | with-exec sh,-c,"echo last > /out" | terminal'`  
`● Attaching terminal:`  
    `container: Container!`  
    `Container.from(address: "docker.io/library/alpine:latest@sha256:21dc6063fd678b478f57c0e13f47560d0ea4eeba26dfc947b2a4f81f686b9f45"): Container!`  
    `Container.withExec(args: ["sh", "-c", "echo last > /out"]): Container!`

`dagger / $ cat /out`  
`last`

`AsService/Up` \- Since pre-built container images for popular long-running services  (e.g. database, webserver, cache, proxy) often use `ENTRYPOINT+CMD` to execute an `entrypoint.sh` script to perform essential service configuration, for calls to `Up` or `AsService` on a `Container` without `args` and with NO mutations/overrides to the `ENTRYPOINT+CMD`/`defaultArgs`, execute `ENTRYPOINT+CMD` (one or both of `ENTRYPOINT` or `CMD` could be empty) on service start as if `useEntrypoint=True` were set (as in \[1\], \[2\] below).

*Note: If both `ENTRYPOINT+CMD` would run, but they are both empty, we should either allow the natural error to occur.*

*Note: Calling `Up` directly on a `Container` is equivalent to first calling `AsService`  (with no args) and then `Up`, and will use the same `CMD`/`DefaultArgs`.*

\[1\] `dagger shell -c 'container | from postgres | with-exposed-port 5432 | with-env-variable POSTGRES_PASSWORD postgres | as-service | up'`

\[2\] `dagger shell -c 'container | from postgres | with-exposed-port 5432 | with-env-variable POSTGRES_PASSWORD postgres | up'`

Note today in 0.15.1, you need to use:  
\[3\] `dagger shell -c 'container | from postgres | with-exposed-port 5432 | with-env-variable POSTGRES_PASSWORD postgres | as-service --use-entrypoint | up'`

Strangely, this does NOT work in 0.15.1:  
\[4\] `dagger shell -c 'container | from postgres | with-exposed-port 5432 | with-env-variable POSTGRES_PASSWORD postgres | as-service --args docker-entrypoint.sh,postgres | up'`  
Since it seems there is no PATH for /usr/local/bin/docker-entrypoint.sh etc, but this DOES work:  
`dagger shell -c 'container | from postgres | with-exposed-port 5432 | with-env-variable POSTGRES_PASSWORD postgres | with-default-args docker-entrypoint.sh,postgres | as-service | up'`

**Comment by jpadams** [comment]
@shykes Looks like we have consensus on this proposal ☝️ from most of team and I know we discussed it live where I believe you were 👍. Could you drop a comment here to record decision to re-introduce the prior behavior of:
>  For calls to AsService on a Container without args and with NO changes/overrides of the ENTRYPOINT+CMD, Dagger will execute ENTRYPOINT+CMD on service start.
